first cut of adding a dangling impl test. fix #45165

bdhirsh opened the pull request:
Operators in PyTorch are registered to the Dispatcher with `def()` and `impl()` calls. There are multiple variations of each function, but in general `def()` is used to register the schema of the op while `impl()` maps the op to a particular kernel. 

Ops should not be able to be `impl()`'d without getting `def()`'d- however, all operator registration happens at static initialization time. Because of the fact that static initialization order is undefined, we cannot detect any "dangling impl's" during static initialization. This PR creates a unit test that checks for and errors out on dangling impl's at runtime.

There were a few ops that actually failed the test, so those are fixed here as well (more details in the comments).

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46484 first cut of adding a dangling impl test. fix #45165**

Differential Revision: [D24392625](https://our.internmc.facebook.com/intern/diff/D24392625)

bdhirsh committed:
first cut of adding a dangling impl test. fix #45165

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -2241,6 +2241,22 @@ TEST(NewOperatorRegistrationTest, testDelayedListener) {
+      }
+      error_str += "\n";
+  }
+  ASSERT_EQ(danglingImpls.size(), 0) << error_str;
```
This test is actually failing already on 13 operators. I'm going to dig more into why / whether or not they're real errors
![image](https://user-images.githubusercontent.com/16311747/96295781-31e08c80-0fbc-11eb-87f6-ee77e2adea08.png)

ezyang replied:
Well, good thing you're doing this PR 👍

bdhirsh committed:
Update on "first cut of adding a dangling impl test. fix #45165"




[ghstack-poisoned]

bdhirsh committed:
Update on "first cut of adding a dangling impl test. fix #45165"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/core/NamedRegistrations.cpp:
```diff
@@ -113,10 +113,6 @@ TORCH_LIBRARY_IMPL(aten, Named, m) {
   m.impl("conj.out", CppFunction::makeFallthrough());
   m.impl("contiguous", CppFunction::makeFallthrough());
   m.impl("copy_", CppFunction::makeFallthrough());
-  m.impl("copy_imag", CppFunction::makeFallthrough());
```
These 5 fn's all don't exist in native_functions.yaml. I also found this PR implying that copy_[real|imag] were removed at some point: https://github.com/pytorch/pytorch/pull/39065

ezyang replied:
Yes, removing these looks super harmless.

bdhirsh commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -82,13 +82,16 @@ TORCH_LIBRARY(quantized, m) {
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv3d_unpack(__torch__.torch.classes.quantized.Conv3dPackedParamsBase packed_weights) -> (Tensor unpacked_weights, Tensor? B_origin)"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv2d_stride(__torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weights) -> int[]"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv2d_padding(__torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weights) -> int[]"));
+  m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv2d_output_padding(__torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weights) -> int[]"));
```
Not exactly sure where these missing quantized fn's were meant to be used, but I don't see any usage of them directly in pytorch source (specifically the 3 `quantized` namespace fn's that I added).

There were also 5 fn def's missing from the `_quantized` namespace, but judging from [this comment](https://github.com/pytorch/pytorch/blob/0c5cd8c2b9cdf473e30bbb1b49ca80ed442813df/aten/src/ATen/native/quantized/library.cpp#L165), the `_quantized` namespace has to do with mobile and will be removed soon.

ezyang replied:
cc @jerryzh168 But adding the schemas for these seems like the right thing to do

jerryzh168 replied:
these are used when there are backends that does not implement the same interface for quantized::conv2d/quantized::conv_transpose_2d and need to extract parameters from packed_weights. e.g. glow.

bdhirsh committed:
Update on "first cut of adding a dangling impl test. fix #45165"


Differential Revision: [D24392625](https://our.internmc.facebook.com/intern/diff/D24392625)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -211,6 +211,26 @@ class CAFFE2_API Dispatcher final {
+  /**
+   * For testing purposes.
+   * Returns a list of all operators that were created through calls to registerImpl(),
+   * without any corresponding calls to registerDef(). This is almost certainly a bug,
```
"After static initialization is done, this is almost certainly a bug"

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -211,6 +211,26 @@ class CAFFE2_API Dispatcher final {
+   *
+   *
+   * This function should be called after static initialization is finished, so that
+   * all registrations with the dispatcher have been finalized.
```
It might be a little clearer to say why this function shouldn't be called during static initialization (and in particular, why we can't enforce this invariant "as we go" registering things)

ezyang reviewed (approved):
Good work. Just need to kill the rest of the errors, I suppose?

bdhirsh committed:
Update on "first cut of adding a dangling impl test. fix #45165"


Differential Revision: [D24392625](https://our.internmc.facebook.com/intern/diff/D24392625)

[ghstack-poisoned]
